client/post-editor/controller.js

```diff
@@ -172,20 +173,40 @@ async function maybeCalypsoifyGutenberg( context, next ) {
 
 	const state = context.store.getState();
 	const siteId = getSelectedSiteId( state );
-	const postType = determinePostType( context );
-	const postId = getPostID( context );
 
 	if (
-		( isCalypsoifyGutenbergEnabled( state, siteId ) ||
-			isGutenlypsoEnabled( state, siteId ) ||
-			isEnabled( 'jetpack/gutenframe' ) ) &&
-		'gutenberg' === getSelectedEditor( state, siteId )
+		! isCalypsoifyGutenbergEnabled( state, siteId ) &&
+		! isGutenlypsoEnabled( state, siteId ) &&
+		! isEnabled( 'jetpack/gutenframe' )
 	) {
+		return next();
+	}
```

Comment:
at this point it seems like we've successfully collapsed the logic into fewer separate concepts, but I'm curious now why we pulled out `'gutenberg' === getSelectedEditor(…)` as it seems like that check is part of what decides to redirect or not.

(note that in this example I've applied [DeMorgan's law](https://en.wikipedia.org/wiki/De_Morgan%27s_laws) to invert the conditionals and make them seem like "if none of these conditions hold then redirect" instead of "if all of these conditions fail to hold" just because in my head the `! ( A || B )` seems clearer as I'm thinking out loud here.

```js
/**
 * page.js middleware to keep the browser URL in sync
 * with state of the block editor in its different environments
 */
async function updateUrlForEditor( context, next ) {
	const needsReplacementUrl = (
		   isCalypsoifyGutenbergEnabled( state, siteId )      // match route of iframe with custom stylesheet to `wp-admin`
		|| isGutenlypsoEnabled( state, siteId )               // manage route state for integrated editor-as-npm module
		|| isEnabled( 'jetpack/gutenframe' )                  // match route for iframed editor on remote site
		|| 'gutenberg' === getSelectedEditor( state, siteId ) // the classic editor handles its own URLs or something
	);

	if ( ! needsReplacementUrl ) {
		return next();
	}

	const postType = determinePostType( context );
	const postId = getPostID( context );

	// without required support in Jetpack we have to
	// fall back to the Calypsoify flow. newer versions
	// support loading the iframe
	// @see `iframe_nonce_preview`
	if ( false === isJetpackMinimumVersion( … ) ) {
		const siteAdminUrl = getSiteAdminUrl( state, siteId );
		const path = postId
			? `post.php?post=${ postId }&action=edit`
			: `post-new.php?post_type=${ postType }`;

		…

		return window.location.replace( addQueryArgs( context.query, url ) );
	}

	const url = getEditorUrl( state, siteId, postId, postType );

	return window.location.replace( addQueryArgs( context.query, url ) );
}

…

updateUrlForEditor( context, next );
```

What have I done?
 - flattened out the logical nesting to keep the indentation (and branching) shallow. specifically I'm trying to avoid the ambiguity introduced in line 202 of the patch when we have a nested `if` but no nested `else`. was it an oversight? was it intentional?
 - renamed some things to make clearer matches to what our intention is. since `window.location.replace()` isn't triggering page navigation I got the impression that the long-wondered-about answer is that this function is managing the location bar's URL to match things that are happening inside the editor, or at least it loads them on editor init.
 - added a basic DocBlock comment to clarify the purpose here for the next dev to come around to it
 - separated some combined logic into multiple steps in order to give names to conditions and remove negation where it was confusing to me

Reply:
Thanks for your feedback!

> I'm curious now why we pulled out `'gutenberg' === getSelectedEditor(…)` as it seems like that check is part of what decides to redirect or not.

It's the common denominator here. We only redirect sites that have Gutenberg as their editor and an outdated Jetpack version.

Currently sites have to use Gutenberg AND have one of the three flags enabled (calypsoifyGutenberg, Gutenlypso, or jetpack/gutenlypso (In your example it's an `||`). So by pulling out the flags check, we can a) bail early and b) remove it cleanly once launched.

> I'm trying to avoid the ambiguity introduced in line 202 of the patch when we have a nested if but no nested else. 

I'm happy to invert the selected editor check and bail early for better readability.

> window.location.replace() isn't triggering page navigation

Based on my testing it _does_ trigger a page load, according to [MDN web docs](https://developer.mozilla.org/en-US/docs/Web/API/Location/replace) however, that location change doesn't get added to history.

Reply:
> Based on my testing it does trigger a page load, according to MDN web docs however, that location change doesn't get added to history.

good call. I should have known that but was confused when I was reading it. thanks